### PR TITLE
Improve SkipNBytes to RandomInputStream 

### LINKED ISSUE #14512: gfile random access file seek implementation inappropriate
the call stack for tensorflow.python.platform.gfile.GFile.seek function is as follows:
https://github.com/tensorflow/tensorflow/blob/294442996b2aeff00b1bfdc7e7169f7cb35bbf3d/tensorflow/python/lib/io/file_io.py#L118
https://github.com/tensorflow/tensorflow/blob/18f36927160d05b941c056f10dc7f9aecaa05e23/tensorflow/core/lib/io/buffered_inputstream.cc#L153
https://github.com/tensorflow/tensorflow/blob/18f36927160d05b941c056f10dc7f9aecaa05e23/tensorflow/core/lib/io/buffered_inputstream.cc#L126
https://github.com/tensorflow/tensorflow/blob/d44d271c9da4d244ce4b2ffaf808adbe4cff759d/tensorflow/core/lib/io/random_inputstream.h#L27
https://github.com/tensorflow/tensorflow/blob/d44d271c9da4d244ce4b2ffaf808adbe4cff759d/tensorflow/core/lib/io/inputstream_interface.cc#L27

The implementation of SkipNBytes for random access file is read out N bytes. It's not efficient if we seek through one huge file. I think we should override the implementation for random access file.

I tried seek API for one large hdfs file, it took more than one hour to finish.

### PR DESCRIPTION
**Pull request by yongtang**
This fix tries to address the issue raised in #14512 where SkipNBytes will takes a long time. In existing implementation, it needs to read through the file even though it is possible to random access the file itself.

This fix convert SkipNBytes in InputStreamInterface to pure virtual so that it is possible to have different implementations based on the underlying file type (seekable vs. no seekable)

This fix then have a specialized implementation in `RandomInputStream` so that
1. The `SkipNBytes` will first try to read 1 bytes from `expected_pos - 1`, if completed, then return.
2. If an out of range error returned from last step, then we don't know how far away we are from the EOF. We are forced to read through the file from the current pos all the way to `expected_pos`. This is similar to existing implementation.

Note: Because `RandomAccessFile` only exposes `Read` (with start offset), not `SizeOfFile`, it is not possible to just move the position marker as we won't know where is the EOF.

One potential improvement could be to expose a `SizeOfFile` in `RandomAccessFile`. That means adding additional changes. It could be added if it is agreed.

This fix fixes #14512.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

### CODE REVIEW

**Comment by martinwicke** [comment]
@rohan100jain can you take a look?

**Comment by drpngx** [comment]
@rohan100jain any luck with this?

**Comment by tensorflowbutler** [comment]
Nagging Assignee: It has been 14 days with no activity and this issue has an assignee. Please update the label and/or status accordingly.

**Comment by rmlarsen** [comment]
@rohan100jain Please take a look.

File: tensorflow/core/lib/io/inputstream_interface.h [hunk 1]
```diff
@@ -38,7 +38,7 @@ class InputStreamInterface {
   // Typical return codes:
   //  * OK - in case of success.
   //  * OUT_OF_RANGE - not enough bytes remaining before end of file.
-  virtual Status SkipNBytes(int64 bytes_to_skip);
+  virtual Status SkipNBytes(int64 bytes_to_skip) = 0;
```
Comment (rohan100jain) [inline tensorflow/core/lib/io/inputstream_interface.h]: I don't understand the need to make this pure virtual... just keep it virtual and have RandomInputStream override with a more efficient implementation. Then you don't have to copy the implementations in the other subclasses.

**Review by rohan100jain — CHANGES_REQUESTED** [review]

**Comment by yongtang** [comment]
Thanks @rohan100jain for the review. The PR has been updated. Please take a look.

**Review by rohan100jain — APPROVED** [review]

**Pull request merged**